tests/pgle_test.py

```diff
@@ -347,5 +348,145 @@ def f(x, y):
+        self.assertNotEqual(non_pgle_f_files, pgle_and_non_pgle_f_files)
+
+        # Remove non-PGLE version of `f` from the cache so a hit in run 3 is
+        # definitely the PGLE version
+        for non_pgle_file in non_pgle_f_files:
+          os.remove(os.path.join(cache_dir, non_pgle_file))
+
+        # Run 3: put a non-PGLE version of `g` in the cache
+        with jtu.count_cached_compilation_cache_miss() as cache_miss_count:
+          g(x)
+        self.assertGreater(cache_miss_count[0], 0)
+
+      api.clear_caches()
+      pjit._pgle_profiler_dict.clear()
+
+      # 2. read from the persistent cache with PGLE disabled-but-expected
+      with (config.enable_compilation_cache(True),
+            config.raise_persistent_cache_errors(True),
+            config.persistent_cache_min_entry_size_bytes(0),
+            config.persistent_cache_min_compile_time_secs(0),
+            config.compilation_cache_expect_pgle(True)):
+        # Run 4 (simulating run 1 in a new process) should pick up the PGLE-optimised
+        # cache entry, even though PGLE is not enabled
+        cache_hit = 0
+        def check_if_cache_hit(event):
+          nonlocal cache_hit
+          if event == '/jax/compilation_cache/cache_hits':
+            cache_hit += 1
+
+        monitoring.register_event_listener(check_if_cache_hit)
+        f(x)
+        monitoring._unregister_event_listener_by_callback(check_if_cache_hit)
+        self.assertGreater(cache_hit, 0)
+
+        # Run 5: `g` was only executed once and did not get re-compiled with PGLE, so
+        # executing it with compilation_cache_expect_pgle will raise a warning and a
+        # cache *hit*, because the non-PGLE version will be loaded
+        with warnings.catch_warnings(record=True) as w:
+          warnings.simplefilter("always")
+          cache_hit = 0
+          monitoring.register_event_listener(check_if_cache_hit)
+          g(x)
+          monitoring._unregister_event_listener_by_callback(check_if_cache_hit)
+          self.assertEqual(cache_hit, 1)
+          if len(w) != 1:
+            print("Warnings:", [str(w_) for w_ in w], flush=True)
+          self.assertLen(w, 1)
+          self.assertIn("PERSISTENT CACHE MISS for PGLE-optimized jit_g despite non-PGLE hit", str(w[0].message))
+
+        # Run 6: `h` was not executed during step 1, which populated the cache, so
+        # executing it now and triggering a cache write will emit a warning
+        with warnings.catch_warnings(record=True) as w:
+          warnings.simplefilter("always")
+          with jtu.count_cached_compilation_cache_miss() as cache_miss_count:
+            h(x)
+          self.assertGreater(cache_miss_count[0], 0)
+          if len(w) != 1:
+            print("Warnings:", [str(w_) for w_ in w], flush=True)
+          self.assertLen(w, 1)
+          self.assertIn("PERSISTENT CACHE WRITE with key jit_h-", str(w[0].message))
+
+  # Test must run last; XLA does not recover from CUPTI errors
```

Comment:
You cannot easily guarantee a test runs last in our test suite, I'm afraid. The only possible way to guarantee this would be to run the test in a subprocess.